File: _experimental/_erlang/words/src/numerical.erl
-module(numerical).

-export([fibonacci/1, factorial/1, gcd/2, lcm/2]).

fibonacci(N) ->
  if
    N < 0 ->
      -1;
    N < 2 ->
      N;
    N >1 ->
      fibonacci_loop(N, 0, 1)
  end.

fibonacci_loop(N, Acc2, Acc1) ->
  if
    N == 2 ->
      Acc1 + Acc2;
    true ->
      fibonacci_loop(N - 1, Acc1, Acc1 + Acc2)
  end.

factorial(N) ->
  if
    N < 0 ->
      -1;
    true ->
      factorial_loop(N, 1)
  end.

factorial_loop(N, Acc) ->
  if
    N == 1 ->
      Acc;
    true ->
      factorial_loop(N - 1, N * Acc)
  end.

gcd(A, B) -> 
  if
    B == 0 ->
      A;
    true ->
      gcd(B, A rem B)
  end.

lcm(A, B) ->
  (A div gcd(A, B)) * B.

File: _experimental/_python/words/src/numerical.py
def fibonacci(n):
  if n < 0:
    return -1
  elif n < 2:
    return n
  else:
    acc2 = 0
    acc1 = 1
    for _ in range(2, n):
      acc2, acc1 = acc1, acc1 + acc2
    return acc1 + acc2
  
def factorial(n):
  if n < 0:
    return -1
  else:
    acc = 1
    for i in range(2, n + 1):
      acc *= i
    return acc
  
def gcd(a, b):
  while b:
    a, b = b, a % b
  return a

def lcm(a, b):
  return (a // gcd(a, b)) * b